### PR TITLE
Apply <inheritdoc /> to Core

### PR DESCRIPTION
**Pull request by roji**
Part of #18003

Will do Relational and the rest in a separate PR, not to have too much fun in one go.

### CODE REVIEW

**Comment by roji** [comment]
image

File: src/EFCore.Abstractions/ChangeTracking/Internal/ObservableBackedBindingList.cs [hunk 1]
```diff
@@ -249,10 +209,10 @@ private void AddToObservableCollection(T item)
...
 
-        // <summary>
-        // Removes the item from the underlying from observable collection.
-        // </summary>
-        // <param name="item"> The item. </param>
+        /// <summary>
+        /// Removes the item from the underlying from observable collection.
+        /// </summary>
+        /// <param name="item"> The item. </param>
         private void RemoveFromObservableCollection(T item)
```
Comment (smitpatel) [inline src/EFCore.Abstractions/ChangeTracking/Internal/ObservableBackedBindingList.cs:216]: Why private members have docs?

Comment (roji) [reply src/EFCore.Abstractions/ChangeTracking/Internal/ObservableBackedBindingList.cs:216]: We did agree to add docs to private members where it makes sense (i.e. a complicated method that needs clarification on its API), but I agree that these docs are useless... will remove.

Comment (smitpatel) [reply src/EFCore.Abstractions/ChangeTracking/Internal/ObservableBackedBindingList.cs:212]: Same in AddToObservableCollection

File: src/EFCore.Cosmos/Storage/Internal/CosmosTransactionManager.cs [hunk 2]
```diff
@@ -99,12 +99,7 @@ public virtual Transaction EnlistedTransaction
         public virtual void EnlistTransaction(Transaction transaction)
             => throw new NotSupportedException();
 
-        /// <summary>
-        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
-        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
-        ///     any release. You should only use it directly in your code with extreme caution and knowing that
-        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
-        /// </summary>
+        /// <inheritdoc />
```
Comment (sharwell) [inline src/EFCore.Cosmos/Storage/Internal/CosmosTransactionManager.cs:102]: ⚠️ There is nothing here to inherit. [SA1648](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1648.md) may be usable for validation of the PR.

File: src/EFCore.Relational/Infrastructure/RelationalDbContextOptionsBuilder.cs [hunk 3]
```diff
@@ -137,27 +137,17 @@ protected virtual TBuilder WithOption([NotNull] Func<TExtension, TExtension> set
 
         #region Hidden System.Object members
 
-        /// <summary>
-        ///     Returns a string that represents the current object.
-        /// </summary>
-        /// <returns> A string that represents the current object. </returns>
+        /// <inheritdoc />
```
Comment (sharwell) [inline src/EFCore.Relational/Infrastructure/RelationalDbContextOptionsBuilder.cs:140]: 📝  The IDE will automatically treat lack of documentation on `override` members as having an implicit `<inheritdoc />`. Some projects leverage this to remove extraneous lines, while other projects prefer to be explicit.

This comment is only intended to present alternatives. The approach preference is left to your team.

File: src/EFCore/ChangeTracking/Internal/ArrayPropertyValues.cs [hunk 4]
```diff
@@ -118,7 +118,7 @@ public override void SetValues(PropertyValues propertyValues)
         ///     doing so can result in application failures when updating to a new Entity Framework Core release.
         /// </summary>
         public override IReadOnlyList<IProperty> Properties
-            => _properties ?? (_properties = EntityType.GetProperties().ToList());
+            => _properties ??= EntityType.GetProperties().ToList();
```
Comment (sharwell) [inline src/EFCore/ChangeTracking/Internal/ArrayPropertyValues.cs:121]: 📝 Possibly unintentional unrelated change

File: src/EFCore.Cosmos/Storage/Internal/CosmosTransactionManager.cs (hunk 2 above)
Comment (sharwell) [reply src/EFCore.Cosmos/Storage/Internal/CosmosTransactionManager.cs:102]: ➡️ This is my mistake. The documentation is inherited from implemented interface member `IResettableService.ResetState()`.

Note that the documentation does change in this case, but that may be desired.

File: src/EFCore/ChangeTracking/Internal/ChangeDetector.cs [hunk 5]
```diff
@@ -49,30 +49,15 @@ public class ChangeDetector : IChangeDetector
             _loggingOptions = loggingOptions;
         }
 
-        /// <summary>
-        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
-        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
-        ///     any release. You should only use it directly in your code with extreme caution and knowing that
-        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
-        /// </summary>
+        /// <inheritdoc />
```
Comment (sharwell) [inline src/EFCore/ChangeTracking/Internal/ChangeDetector.cs:52]: 📝 This is going to inherit from interface documentation, which means it no longer reports the "internal API" warning. For any case where this is not desirable (i.e. you want to include the "internal API" warning but you don't want to write the whole thing every time, you can take the following approach:

1. Create a helper type:

    ```csharp
    /// <summary>
    ///     ... text of the message goes here ...
    /// </summary>
    internal static class InternalApiSummary
    {
    }
    ```

2. For any case where you want to include this message, use this (works for both types and members):

    ```csharp
    /// <inheritdoc cref="InternalApiSummary" />
    ```

File: src/EFCore/ChangeTracking/Internal/ValueGenerationManager.cs [hunk 6]
```diff
@@ -13,20 +13,7 @@
...
-    ///         doing so can result in application failures when updating to a new Entity Framework Core release.
-    ///     </para>
-    ///     <para>
-    ///         The service lifetime is <see cref="ServiceLifetime.Scoped" />. This means that each
-    ///         <see cref="DbContext" /> instance will use its own instance of this service.
-    ///         The implementation may depend on other services registered with any lifetime.
-    ///         The implementation does not need to be thread-safe.
-    ///     </para>
-    /// </summary>
+    /// <inheritdoc />
```
Comment (sharwell) [inline src/EFCore/ChangeTracking/Internal/ValueGenerationManager.cs:16]: 📝 For a case like this where you want both the inherited documentation and the documentation from `InternalApiSummary`, you do have the option of doing something like this:

```csharp
/// <summary>
/// <inheritdoc />
/// <inheritdoc cref="InternalApiSummary" />
/// </summary>
```

Or if you want to put the "internal API" warning in the remarks section instead of the summary section, you could do this:

```csharp
/// <inheritdoc />
/// <remarks>
/// <inheritdoc cref="InternalApiSummary" path="/summary" />
/// </remarks>
```

Comment (sharwell) [reply src/EFCore/ChangeTracking/Internal/ValueGenerationManager.cs:16]: 💭 I suppose if it was common to use the second of these approaches, you could define an `InternalApiRemarks` class to help:

```
/// <remarks>
/// <inheritdoc cref="InternalApiSummary" path="/summary" />
/// </remarks>
internal static class InternalApiRemarks
{
}
```

Then you could do this to inherit from base plus include the internal API remarks:

```csharp
/// <inheritdoc />
/// <inheritdoc cref="InternalApiRemarks" />
```

File: src/EFCore.Relational/Infrastructure/RelationalDbContextOptionsBuilder.cs (hunk 3 above)
Comment (roji) [reply src/EFCore.Relational/Infrastructure/RelationalDbContextOptionsBuilder.cs:140]: I think not having to say this explicitly would be great... But I was under the impression that docfx needs them... Does anyone have any info on this /cc @smitpatel @bricelam

File: src/EFCore/ChangeTracking/Internal/ArrayPropertyValues.cs (hunk 4 above)
Comment (roji) [reply src/EFCore/ChangeTracking/Internal/ArrayPropertyValues.cs:121]: Unrelated but intentional :)

File: src/EFCore.Relational/Infrastructure/RelationalDbContextOptionsBuilder.cs (hunk 3 above)
Comment (bricelam) [reply src/EFCore.Relational/Infrastructure/RelationalDbContextOptionsBuilder.cs:140]: So long as IntelliSense and our API docs show it, I'm good with whatever. Not sure whether they require it or not.

Comment (roji) [reply src/EFCore.Relational/Infrastructure/RelationalDbContextOptionsBuilder.cs:140]: Right, the question here is whether our API docs (generated with docfx) require the explicit `<inheritdoc />` or not... Does anyone have an answer to that question? Does @smitpatel have some contacts in the doc team we could ask about this?

Comment (smitpatel) [reply src/EFCore.Relational/Infrastructure/RelationalDbContextOptionsBuilder.cs:140]: Query sent. It is async query so I will await for a response.

Comment (roji) [reply src/EFCore.Relational/Infrastructure/RelationalDbContextOptionsBuilder.cs:140]: Thanks @smitpatel, hopefully there is an SLA in place on the receiving side...

File: src/EFCore/ChangeTracking/Internal/ValueGenerationManager.cs (hunk 6 above)
Comment (roji) [reply src/EFCore/ChangeTracking/Internal/ValueGenerationManager.cs:16]: Wow, that's next-level doc wizardry, good to know this is possible... For the last part, is the inherited summary automatically appended to the non-inherited one, and the same for each other top-level doc section?

In any case, I'm a bit reluctant to start combining the "internal API" blurb with actual docs, since we may be dropping pubternal for 6.0 and then that work would be lost.

File: src/EFCore/ChangeTracking/Internal/ChangeDetector.cs (hunk 5 above)
Comment (roji) [reply src/EFCore/ChangeTracking/Internal/ChangeDetector.cs:52]: I'll definitely do that.

File: src/EFCore.Relational/Infrastructure/RelationalDbContextOptionsBuilder.cs (hunk 3 above)
Comment (smitpatel) [reply src/EFCore.Relational/Infrastructure/RelationalDbContextOptionsBuilder.cs:140]: The person who knows this is on vacation. Expect delays.

Comment (roji) [reply src/EFCore.Relational/Infrastructure/RelationalDbContextOptionsBuilder.cs:140]: An idea: do we know if the API docs just use the same docfx that's available publicly (https://github.com/dotnet/docfx)? If so, I use that for the Npgsql API docs and can possibly test there.

Comment (smitpatel) [reply src/EFCore.Relational/Infrastructure/RelationalDbContextOptionsBuilder.cs:140]: To answer you question, I don't know. The job which produces the docs may be using that tool but there could be .NET specific customizations on top of that. (there are json files in the API docs repo).

Rather than trying to do additional work and figure out a lot of unknowns, is there harm in just waiting for doc team to respond? What is rush to get this merged?

Comment (roji) [reply src/EFCore.Relational/Infrastructure/RelationalDbContextOptionsBuilder.cs:140]: No harm or rush. It would just be useful to know whether the publicly-available docfx is the same one being used for our docs, this way we easily test and see what it supports and what it doesn't.

**Comment by smitpatel** [comment]
Another blocked label. Doc team said they don't support <inheritdoc /> in API generation.

**Comment by roji** [comment]
> Another blocked label. Doc team said they don't support in API generation.

What exactly isn't support? If you mean they don't support `<inheridoc />` at all, then we have a problem because we're already using it in many places. If it's supported but not implicitly assumed, then this PR as-is doesn't do that (so doesn't need to be blocked?)

**Comment by ajcvickers** [comment]
@roji It's mainly blocked to decide whether we really want to replace the intentionally in-your-face pubternal boilerplate message with another boilerplate message that is not in-your-face at all.

**Comment by roji** [comment]
@ajcvickers do you mean we're considering preserving the internal boilerplate duplicated everywhere in the *source code (as opposed to in the generated docs, which this PR isn't supposed to change)?

If so, maybe this whole discussion should be deferred until we decide what we want to do with pubternals.

**Comment by ajcvickers** [comment]
Yes, and yes.

**Comment by ajcvickers** [comment]
@roji Docs team has confirmed that <inheritdoc /> is not actually supported, and we have verified that this is the case, so I think this will need to be closed. 😭

**Comment by roji** [comment]
Very odd, [a quick search of issues around inheritdoc](https://github.com/dotnet/docfx/issues?q=is%3Aissue+is%3Aopen+inheritdoc) seems to show that docfx does support it, at least to some extent. Npgsql also uses docfx, and it does seem to handle at least simple cases of inheritdoc inside the codebase (e.g. [these API docs](http://www.npgsql.org/doc/api/Npgsql.TypeHandlers.BoolHandler.html#Npgsql_TypeHandlers_BoolHandler_Read_Npgsql_NpgsqlReadBuffer_System_Int32_Npgsql_BackendMessages_FieldDescription_) are generated from [this code](https://github.com/npgsql/npgsql/blob/8a479081f707784b5040747b23102c3d6371b9d3/src/Npgsql/TypeHandlers/BoolHandler.cs#L27)).

I wonder if we're getting a general "not completely implemented" answer whereas what we need may work. Or maybe the MS docs use some custom version without the support, who knows.

**Comment by ajcvickers** [comment]
@roji Smit posted example in triage Teams chat of how it doesn't work, in addition to the docs team saying it isn't supported.

**Comment by roji** [comment]
Should we just close it? It's not like we expect it to work anytime soon...

**Comment by smitpatel** [comment]
Doc team has not prioritized adding this feature yet. So we should just close the PR (also looking at the stale PR graphs we got in emails).
Save the commit if you want to preserve to use as starting point later.

**Comment by roji** [comment]
Saved in https://github.com/roji/efcore/tree/InheritALLTheDocs
